### PR TITLE
Support ISerializable in ReflectionOnly mode.

### LINKED ISSUE #13071: ReflectionOnly Mode not Supporting ISerializable
PR https://github.com/dotnet/corefx/pull/13068 enabled DCS/DCJS to serialize ISerializable types. We still need to make changes to enable reflection based serialization to support ISerializable. The following test in System.Runtime.Serialization.Xml.ReflectionOnly.Tests are failing,

```
DataContractJsonSerializerTests.DCJS_MyISerializableType
DataContractSerializerTests.DCS_ExceptionObject
DataContractSerializerTests.DCS_InnerExceptionMesageWithSpecialChars
DataContractSerializerTests.DCS_ExceptionMesageWithSpecialChars
DataContractSerializerTests.DCS_ArgumentExceptionObject
DataContractSerializerTests.DCS_MyISerializableType
```

To fix this, take a look at PR https://github.com/dotnet/corefx/pull/13068 first, and changes probably need to be made in the following files,
```
ReflectionXmlFormatReader.cs
ReflectionXmlFormatWriter.cs
ReflectionJsonFormatReader.cs
ReflectionJsonFormatWriter.cs
```

### PR DESCRIPTION
**Pull request by shmao**
Enable DataContractSerializer and DataContractJsonSerializer to support ISerializable types in ReflectionOnly mode.

Fix #13071

### CODE REVIEW

**Comment by shmao** [comment]
@dotnet-bot test Innerloop OSX 

The following test failed.
> System.Net.Security.Tests.SslStreamSystemDefaultTest.SslStream_DefaultTlsConfigurationSync_Ok

**Comment by shmao** [comment]
@dotnet-bot test Innerloop OSX Debug Build and Test, Innerloop OSX Release Build and Test

File: src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/ReflectionReader.cs [hunk 1]
```diff
@@ -279,6 +287,16 @@ private object ReadItemOfPrimitiveType(XmlReaderDelegator xmlReader, XmlObjectSe
             return value;
         }
 
+        private static object ReadISerializable(XmlReaderDelegator xmlReader, XmlObjectSerializerReadContext context, ClassDataContract classContract)
+        {
+            object obj;
+            SerializationInfo serializationInfo = context.ReadSerializationInfo(xmlReader, classContract.UnderlyingType);
+            StreamingContext streamingContext = context.GetStreamingContext();
+            ConstructorInfo iSerializableConstructor = classContract.GetISerializableConstructor();
+            obj = iSerializableConstructor.Invoke(new object[] { serializationInfo, streamingContext });
```
Comment (mconnew) [inline src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/ReflectionReader.cs]: Could a stronger typed delegate be returned instead? All ISerializable constructors have the same parameters.

File: src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/ReflectionClassWriter.cs [hunk 2]
```diff
@@ -21,7 +21,16 @@ public void ReflectionWriteClass(XmlWriterDelegator xmlWriter, object obj, XmlOb
         {
             InvokeOnSerializing(obj, context, classContract);
             obj = ResolveAdapterType(obj, classContract);
-            ReflectionWriteMembers(xmlWriter, obj, context, classContract, classContract, 0 /*childElementIndex*/, memberNames);
+            var iSerializableObj = obj as ISerializable;
```
Comment (mconnew) [inline src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/ReflectionClassWriter.cs]: Yes, checking a bool is going to be cheaper than a cast which will need to traverse the type hierarchy to look for ISerializable in the type and parents.

Comment (shmao) [inline src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/ReflectionClassWriter.cs]: Fixed this.

File: src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/ReflectionReader.cs [hunk 3]
```diff
@@ -33,7 +33,15 @@ public object ReflectionReadClass(XmlReaderDelegator xmlReader, XmlObjectSeriali
             context.AddNewObject(obj);
             InvokeOnDeserializing(context, classContract, obj);
 
-            ReflectionReadMembers(xmlReader, context, memberNames, memberNamespaces, classContract, ref obj);
+            if (classContract.IsISerializable)
+            {
+                obj = ReadISerializable(xmlReader, context, classContract);
```
Comment (shmao) [inline src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/ReflectionReader.cs]: I opened https://github.com/dotnet/corefx/issues/14651 for tracking potential issues with reading ISerializable.

File: src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/ReflectionReader.cs (hunk 1 above)
Comment (shmao) [inline src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/ReflectionReader.cs]: I opened https://github.com/dotnet/corefx/issues/14693 for tracking this.

**Pull request merged**